### PR TITLE
Add ClickBench queries to DataFusion benchmark runner

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/arrow-datafusion/pull/7054~

# Which issue does this PR close?
closes https://github.com/apache/arrow-datafusion/issues/6994
closes https://github.com/apache/arrow-datafusion/issues/6128

# Rationale for this change
see https://github.com/apache/arrow-datafusion/issues/6994 -- tldr is to optimize clickbench queries it needs to be easier to run them

# What changes are included in this PR?
1. Add new `dfbench clickbench` command to run ClickBench queries
2. Update `bench.sh` to run clickbench queries
3. Update `benchmarks/README.md` -- see rendered version https://github.com/alamb/arrow-datafusion/tree/alamb/clickbench_runner/benchmarks

# Are these changes tested?
I tested them manually

Run clickbench q1 directly (e.g. for profiling):

```shell
cargo run  --bin dfbench -- clickbench --query 1
Running benchmarks with the following options: RunOpt { query: Some(1), common: CommonOpt { iterations: 3, partitions: 2, batch_size: 8192 }, path: "benchmarks/data/hits.parquet", queries_path: "benchmarks/queries/clickbench/queries.sql", output_path: None }
Q1: SELECT COUNT(*) FROM hits;
Query 1 iteration 0 took 305.7 ms and returned 1 rows
Query 1 iteration 1 took 13.6 ms and returned 1 rows
Query 1 iteration 2 took 13.6 ms and returned 1 rows
```

run with hits_partitioned (100 parquet files):
```shell
cargo run  --bin dfbench -- clickbench --query 1 --path=benchmarks/data/hits_partitioned
```

Run with bench.sh:

```shell
./bench.sh run clickbench_1
```

See help
```shell
cargo run  --bin dfbench  -- clickbench --help

dfbench-clickbench 27.0.0
Run the clickbench benchmark

The ClickBench[1] benchmarks are widely cited in the industry and
focus on grouping / aggregation / filtering. This runner uses the
scripts and queries from [2].

[1]: https://github.com/ClickHouse/ClickBench
[2]: https://github.com/ClickHouse/ClickBench/tree/main/datafusion

USAGE:
    dfbench clickbench [OPTIONS]

FLAGS:
    -h, --help       
            Prints help information

    -V, --version    
            Prints version information


OPTIONS:
    -s, --batch-size <batch-size>        
            Batch size when reading CSV or Parquet files [default: 8192]

    -i, --iterations <iterations>        
            Number of iterations of each test run [default: 3]

    -o, --output <output-path>           
            If present, write results json here

    -n, --partitions <partitions>        
            Number of partitions to process in parallel [default: 2]

    -p, --path <path>                    
            Path to hits.parquet (single file) or `hits_partitioned` (partitioned, 100 files) [default:
            benchmarks/data/hits.parquet]
    -r, --queries_path <queries-path>    
            Path to queries.sql (single file) [default: benchmarks/queries/clickbench/queries.sql]

    -q, --query <query>                  
            Query number. If not specified, runs all queries

```

# Are there any user-facing changes?

### CODE REVIEW

File: benchmarks/src/options.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::prelude::SessionConfig;
+use structopt::StructOpt;
+
+// Common benchmark options (don't use doc comments otherwise this doc
+// shows up in help files)
+#[derive(Debug, StructOpt, Clone)]
```
Comment (alamb) [inline benchmarks/src/options.rs:23]: refactored common options into `CommonOpt`

File: benchmarks/src/lib.rs [hunk 2]
```diff
@@ -16,27 +16,10 @@
...
-
-pub mod run;
+pub mod clickbench;
 pub mod tpch;
 
+mod options;
+mod run;
+pub use options::CommonOpt;
 pub use run::{BenchQuery, BenchmarkRun};
-
```
Comment (alamb) [inline benchmarks/src/lib.rs:26]: this was dead code I accidentally introduced in https://github.com/apache/arrow-datafusion/pull/7054

The actual entrypoint for the dfbench binary is in https://github.com/apache/arrow-datafusion/blob/main/benchmarks/src/bin/dfbench.rs

File: benchmarks/src/tpch/run.rs [hunk 3]
```diff
@@ -42,8 +42,17 @@ use structopt::StructOpt;
 
 use super::{get_tbl_tpch_table_schema, get_tpch_table_schema, TPCH_TABLES};
 
-/// Run the tpch benchmark
+/// Run the tpch benchmark.
+///
+/// This benchmarks is derived from the [TPC-H][1] version
```
Comment (alamb) [inline benchmarks/src/tpch/run.rs:47]: I moved the details of what the benchmark was doing into the binary from the README, which I think is better as it is closer to the code, but I don't feel strongly about this and would welcome other opinions on the matter

**Comment by alamb** [comment]
@tustvold or @Dandandan  do you have time to review this PR (I hope to use this benchmark runner to drive / test further groupby peformance improvements)

File: benchmarks/src/clickbench.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                DataFusionError::Context(
+                    format!("Registering 'hits' as {path}"),
+                    Box::new(e),
+                )
+            })
+    }
+
+    /// Returns the text of query `query_id`
+    fn get_query(&self, query_id: usize) -> Result<String> {
+        if query_id == 0 || query_id > 43 {
```
Comment (Dandandan) [inline benchmarks/src/clickbench.rs:125]: ClickBench numbers the queries 0-42:

https://benchmark.clickhouse.com/

**Comment by Dandandan** [comment]
One comment about the clickbench numbers, the rest looks good to me! Thanks for driving this forward

File: benchmarks/src/clickbench.rs (hunk 4 above)
Comment (alamb) [reply benchmarks/src/clickbench.rs]: Good catch -- fixed ae4ce7d

**Review by Dandandan — APPROVED** [review]

**Pull request merged**